### PR TITLE
SSA caching

### PR DESCRIPTION
**Pull request by rafaelsamenezes**
This PR is a basis for the full caching framework. This adds:

1. SSA algorithms, which work similarly to the goto algorithms. This will help develop "passes" over SSA
2. A basic CRC caching. Checks if a guard/cond pair was already seen and replace the assertion with a trivial case. I am not sure if it is better to replace the `cond` with 1 or the `guard` with 0.

### CODE REVIEW

File: src/esbmc/bmc.cpp
Comment (lucasccordeiro) [inline src/esbmc/bmc.cpp]: Improve this comment :-)

File: src/util/algorithms.cpp
Comment (lucasccordeiro) [inline src/util/algorithms.cpp]: Can you please add some comments to explain this piece of code?

File: src/util/cache.h
Comment (lucasccordeiro) [inline src/util/cache.h]: Add some comments to document this class.

**Review by lucasccordeiro — CHANGES_REQUESTED** [review]
@rafaelsamenezes: Could you please run this PR over the SV-COMP benchmarks? It would be good to see the results with and without the cache system.

File: src/esbmc/bmc.cpp
Comment (lucasccordeiro) [inline src/esbmc/bmc.cpp]: "options" -> "option".

**Review by lucasccordeiro — APPROVED** [review]
LGTM.

I would like to see the SV-COMP results before we merge this PR.

**Review by fbrausse — COMMENTED** [review]
That's a nice idea. I have a few concerns about the current implementation:
1. Most importantly: hashing is not injective! There can be (unlikely, but possibly) hash collisions. Replacing a condition solely based on equality of hashes is dangerous. Can we include a check for equality of the actual terms?
2. What are all these new `static` local variables doing there? Are they necessary? Why are they necessarily unique?
3. Using a `std::map` with an enum as key is inefficient. As are all these virtual calls in `ssa_algorithm::run`. Clearly, this class is meant to be subclassed and make these derived implementations shorter. But it does so at the expense of runtime. Is it really that much more cumbersome to put the loop into the subclass? The number of SSA-steps in particular can be very large.

**Comment by rafaelsamenezes** [comment]
> Most importantly: hashing is not injective! There can be (unlikely, but possibly) hash collisions. Replacing a condition solely based on equality of hashes is dangerous. Can we include a check for equality of the actual terms?

Nice catch, I will add it.

> What are all these new static local variables doing there? Are they necessary? Why are they necessarily unique?

One of those was a mistake. Right now I added only one static variable which is the cache itself. I am unsure whether I should keep it static or move it into config.

> Using a std::map with an enum as key is inefficient. 

I removed the map and just added a plain switch. 

> As are all these virtual calls in ssa_algorithm::run. Clearly, this class is meant to be subclassed and make these derived implementations shorter. But it does so at the expense of runtime. Is it really that much more cumbersome to put the loop into the subclass? The number of SSA-steps in particular can be very large.

@fbrausse I think that the runtime impact won't be that impactful. For instance, this was inspired by LLVM passes which also rely on virtual methods for the code transformation/analyses. I would argue the other way around, if an algorithm has to be more performant then we can just do it without the algorithm interface.

Edit: Using std::function members and initialization though the constructor would be better?

**Comment by fbrausse** [comment]
That's not what I meant :) `do_crc()` computes a hash of the type2t or expr2t. Equal `do_crc()` hashes do not imply equal type2t or expr2t. Couldn't you just store the condition and guard in a unordered_set? Might need to specialize `std::hash` for type2t and expr2t...

The `static auto cache` in `bmct::run_thread()` is pretty confusing.to me, as is part of the rest of the API you're introducing.
I would've expected
- `cache` to be local state (a member) of `class crc_assert_cache`, not just a reference to an externally managed set, and
- `ssa_algorithm::run()` to be invoked with `SSA_steps` as parameter - that's what it should run on, isn't it?

If you really want to keep `cache` external to the "pass", wouldn't it make more sense as a member of `class bmct`?

As a side note: To me `do_crc` is a completely misleading name:
a) it caches its result, doesn't "do" anything
b) it recursively computes a hash, it has nothing to do with CRCs or checksums in general.

**Comment by rafaelsamenezes** [comment]
> That's not what I meant :) do_crc() computes a hash of the type2t or expr2t. Equal do_crc() hashes do not imply equal type2t or expr2t. 

I see what you mean now. 

> Couldn't you just store the condition and guard in a unordered_set? 

I am not sure if I follow: you mean something such: `std::pair<const &expr2tc, const &expr2tc>`?

> Might need to specialize std::hash for type2t and expr2t

I think we have it somewhere in the code (we do state hashing after all). I will look for it

> cache to be local state (a member) of class crc_assert_cache, not just a reference to an externally managed set,

The thing is, how to pass this member between runs? We need some way to store/load it between `run_bmc` calls. In this case, I was thinking about kind and incremental runs as the same assertions are going to be verified again.

> ssa_algorithm::run() to be invoked with SSA_steps as parameter - that's what it should run on, isn't it?

I would say the run is just to... run it. When I came up with the idea of the algorithm interface I was thinking of being able to create a list of transformations/analyses and iterate over it with the `run` method (similar to a pass manager of LLVM).

> f you really want to keep cache external to the "pass", wouldn't it make more sense as a member of class bmct?

Kinda yes, however, I felt that would affect readability if we had a member of a class that is only used in a specific and optional path.

**Comment by fbrausse** [comment]
> > Couldn't you just store the condition and guard in a unordered_set?
> 
> I am not sure if I follow: you mean something such: `std::pair<const &expr2tc, const &expr2tc>`?

Yes, maybe even without references: `std::unordered_set<std::pair<expr2tc,expr2tc>>`.

> > Might need to specialize std::hash for type2t and expr2t
> 
> I think we have it somewhere in the code (we do state hashing after all). I will look for it

Thanks, I also recall seeing a std::hash-compatible type somewhere.

> > cache to be local state (a member) of class crc_assert_cache, not just a reference to an externally managed set,
> 
> The thing is, how to pass this member between runs? We need some way to store/load it between `run_bmc` calls. In this case, I was thinking about kind and incremental runs as the same assertions are going to be verified again.

Well, you're now in `bmct::run_thread()` creating a temporary, anonymous object of type `crc_assert_cache` only to invoke `run()` and destroy it again. Wouldn't keeping around this `crc_assert_cache` object in `bmct` (or config, depending on whether we want to centralize "state") take care of passing the cache?

> > ssa_algorithm::run() to be invoked with SSA_steps as parameter - that's what it should run on, isn't it?
> 
> I would say the run is just to... run it.

`ssa_algorithm::run()` would seem to disagree by actually only iterating through the SSA-steps :)

> When I came up with the idea of the algorithm interface I was thinking of being able to create a list of transformations/analyses and iterate over it with the `run` method (similar to a pass manager of LLVM).

Disclaimer: personal opinion following, no technical reason, just a suggestion regarding API design. The types (`goto_functionst` and `SSA_stepst`) the 2 specific kinds of algorithms are operating on, respectively, are fixed by the architecture of ESBMC. We could move the `run()` method from the generic base class `algorithm` to the `goto_functions_algorithm` and `ssa_algorithm`, each taking as argument the respective type they work on. Then `algorithm` itself wouldn't be runnable¹, but that's OK again by the design of ESBMC: After symex, all information is transferred from GOTO to SSA. It doesn't make a lot of sense to postpone running a GOTO algorithm after an SSA one. Then we'd have 2 separate kinds of "passes".

¹ Alternatively, `algorithm` could be a class template taking the argument type of `run()` as template parameters.

> > f you really want to keep cache external to the "pass", wouldn't it make more sense as a member of class bmct?
> 
> Kinda yes, however, I felt that would affect readability if we had a member of a class that is only used in a specific and optional path.

In that case, maybe the config is a better place as it already keeps a bunch of optionally set global information around.

File: src/util/algorithms.cpp
Comment (kunjsong01) [inline src/util/algorithms.cpp]: missing default?

File: src/util/cache.cpp
Comment (kunjsong01) [inline src/util/cache.cpp]: should we report an error and abort here?

File: src/util/cache.h
Comment (kunjsong01) [inline src/util/cache.h]: Perhaps a comment to explain the pair?
e.g. // first refers to guard_crc, second refers to cond_crc?

File: src/esbmc/bmc.cpp
Comment (kunjsong01) [inline src/esbmc/bmc.cpp]: Why static?

**Review by kunjsong01 — CHANGES_REQUESTED** [review]

**Review by kunjsong01 — APPROVED** [review]
LGTM, just left some minor comments above.

**Comment by rafaelsamenezes** [comment]
> ssa_algorithm::run() would seem to disagree by actually only iterating through the SSA-steps :)

I could make every runOnX function be abstract, in this way you couldn't invoke ssa_algorithm directly. It will be more verbose for the inherited classes though.

**Comment by rafaelsamenezes** [comment]
> Disclaimer: personal opinion following, no technical reason, just a suggestion regarding API design. The types (goto_functionst and SSA_stepst) the 2 specific kinds of algorithms are operating on, respectively, are fixed by the architecture of ESBMC. We could move the run() method from the generic base class algorithm to the goto_functions_algorithm and ssa_algorithm, each taking as argument the respective type they work on. Then algorithm itself wouldn't be runnable¹, but that's OK again by the design of ESBMC: After symex, all information is transferred from GOTO to SSA. It doesn't make a lot of sense to postpone running a GOTO algorithm after an SSA one. Then we'd have 2 separate kinds of "passes".

@fbrausse Could you elaborate more? By what I understood you mean removing the `run` method from `algorithms`. And, every inheritance would create it is own run, e.g. `class foo_algorithm : public algorithm { bool run(Foo &foo); };`. I don't mind changing this, it would solve the issue of the static variable. So:

1. Does still make sense to have the algorithms class?
2. When/Where should we initialize the algorithms? My guess is at bmc class

**Comment by fbrausse** [comment]
> I could make every runOnX function be abstract, in this way you couldn't invoke ssa_algorithm directly. It will be more verbose for the inherited classes though.

Don't think that's needed.

> Could you elaborate more? By what I understood you mean removing the `run` method from `algorithms`. And, every inheritance would create it is own run, e.g. `class foo_algorithm : public algorithm { bool run(Foo &foo); };`.

Yes, that's exactly what I was thinking. 

> 1. Does still make sense to have the algorithms class?

There would be only `has_sideeffect()` left in `algorithm`. I can't find users of it though it seems potentially necessary information for classifying when to run a pass. Though an algorithm will have a hard time returning anything meaningful here without knowledge of the data it is being run on. In general this info is part of the run itself.

> 2. When/Where should we initialize the algorithms? My guess is at bmc class

I suppose. You mentioned the initial motivation was to have a list of passes to be run. Does this list exist anywhere, yet?

Btw. the slicer could also be an algorithm, though it obtains all its state from the config at the moment....

**Comment by rafaelsamenezes** [comment]
Ok, from @fbrausse suggestions, I think this should be put on hold. First I will create a PR for:

1. Convert `algorithms` into a template class (where `run` will receive the specialization).
2. Adapt the current goto_functions algorithms to use this new flow.
3. At esbmc_parseoptions class, add a list of algorithm<goto_functionst> to be run during `process_goto_program`
4. Create the ssa_step_algorithm and convert the slicer into it, moving the list of ignored symbols into the algorithm itself.
5. At bmct  class, add list of algorithm<ssa_step> (or similar) to be executed at `run_thread`

After that, I will come back to this PR. What do you think?

**Comment by fbrausse** [comment]
That sounds great!

File: src/util/algorithms.cpp
Comment (fbrausse) [reply src/util/algorithms.cpp]: I don't think `default:` is missing: what would you do in there? The way it is now, we'll get a compiler warning when the `ssa_type` enum changes. IMHO that's better than either aborting (making the compiled code slower) or just `break`, which only hides the warning.

File: src/util/cache.cpp
Comment (rafaelsamenezes) [reply src/util/cache.cpp]: I don't think it is an error. It just means that we shouldn't replace the assertion.

File: src/esbmc/bmc.cpp
Comment (rafaelsamenezes) [reply src/esbmc/bmc.cpp]: I moved it into config

File: src/util/cache.cpp
Comment (fbrausse) [inline src/util/cache.cpp]: `emplace` only if not found is duplicate work. You can use `auto [it,ins] = assert_set.emplace(pair);`, then the boolean `ins` will hold the info whether `pair` was in this set before.

File: src/util/cache.cpp
Comment (fbrausse) [inline src/util/cache.cpp]: As discussed [earlier](https://github.com/esbmc/esbmc/pull/827#issuecomment-1198354788), could you put the expressions themselves into the `assert_set` instead of just the hashes? The `std::unordered_set` will take care of hash collisions.

File: src/util/config.h
Comment (fbrausse) [inline src/util/config.h]: I think you want std::unordered_set as it is hash-based, std::set relies on the order.

**Review by fbrausse — CHANGES_REQUESTED** [review]

**Comment by rafaelsamenezes** [comment]
I will still work on this PR but the initial results look good:

300s, Master
```
Statistics:          15648 Files
  correct:            8456
    correct true:     5128
    correct false:    3328
  incorrect:            54
    incorrect true:     25
    incorrect false:    29
  unknown:            7138
  Score:             12320 (max: 25567)
```

300s, this PR
```
Statistics:          15648 Files
   correct:            8561
     correct true:     5169
     correct false:    3392
   incorrect:            56
     incorrect true:     25
    incorrect false:    31
   unknown:            7031
   Score:             12434 (max: 25567)
```

**Comment by rafaelsamenezes** [comment]
@fbrausse would you mind taking another look? I had to create a new hashing function in order to use `unordered_set`

File: src/util/config.h
Comment (fbrausse) [inline src/util/config.h]: Could you mention the "cache-asserts" option here in order to add some context?

File: src/util/crypto_hash.h
Comment (fbrausse) [inline src/util/crypto_hash.h]: C++20 is already released. It *does* have this function. As a free-standing function. Since the `hash` array is already publicly accessible, there is no need for this function in the `crypto_hash` class. Maybe move it to the unit test case?

File: src/util/crypto_hash.h
Comment (fbrausse) [inline src/util/crypto_hash.h]: You're throwing away bits of the hash through this shift. As `crypto_hash` is already supposed to be a crypto-strength hash¹, the probability of `hash[0] ^ hash[1]` having any recognizeable pattern is the same as that of `hash[0] ^ (hash[1] << 4)`. Similarly the probability of h0,h1 is the same as that of h1,h0. You're not combining hash values of independent objects where the order matters. I would remove the shift.

¹ I am aware that this is SHA-1, which is broken. Not for hash-maps, though. Still, if that's a problem change the implementation to anything that satisfies the `crypto_hash` promise.

**Review by fbrausse — APPROVED** [review]
Very nice PR, thanks for implementing it! I only have a few minor comments about the hashing.

File: src/util/crypto_hash.h
Comment (rafaelsamenezes) [reply src/util/crypto_hash.h]: I have changed the implementation to use only ^. I will also start another sv-comp run to see if it changes anything.

**Comment by lucasccordeiro** [comment]
@rafaelsamenezes: I have just updated the stats file at https://github.com/esbmc/esbmc/blob/master/scripts/competitions/svcomp/stats-300s.txt. 

Next time, before merging the PR, can I ask you to update this file?

**Pull request merged**
